Update for Shadow DOM

TakayoshiKochi opened the pull request:
See discussion about this change at:
https://github.com/w3c/webcomponents/issues/192

Splits Document.pointerLockElement into Document.pointerLockElement and
ShadowRoot.pointerLockElement and encapsulates shadow tree from
document.

Other than split, replaces document to 'shadow-including document' where
necessary.

Addresses #3.

TakayoshiKochi committed:
Update for Shadow DOM

See discussion about this change at:
https://github.com/w3c/webcomponents/issues/192

Splits Document.pointerLockElement into Document.pointerLockElement and
ShadowRoot.pointerLockElement and encapsulates shadow tree from
document.

Other than split, replaces document to use Shadow DOM-aware vocabulary
like 'shadow-including root' where necessary.

Fixes #3.

TakayoshiKochi commented:
@scheib @domenic @annevk Could you review?

domenic commented on index.html:
```diff
@@ -239,8 +239,9 @@
-          <a href="http://www.whatwg.org/specs/web-apps/current-work/multipage/infrastructure.html#in-a-document">in</a>
+          <p>Pointer lock must succeed only if the <a>target</a>'s
+          <a href="https://dom.spec.whatwg.org/#concept-shadow-including-root">shadow-including root</a> is
+          <a href="https://dom.spec.whatwg.org/#in-a-shadow-including-document">in</a>
```
I think it's not "in", it's just "is"

domenic commented on index.html:
```diff
@@ -254,15 +255,15 @@
           gesture</a>, or pointer lock has not previously been entered for this
-          document, an event generated as a result of an
-          <a>engagement gesture</a> must be received by the document before
+          shadow-including document, an event generated as a result of an
```
I don't believe this is necessary. In general there is no such term as "shadow-including document". It is just "document". You can be "in a shadow-including document" but the "shadow-including" modifies "in", not "document." So any case where there is no "in" seems like a wrong change.

domenic commented on index.html:
```diff
@@ -349,6 +341,24 @@
+           <dfn title="pointerLockElement"></dfn>
+
+           <p>Returns the result of the
+           <a href="https://w3c.github.io/webcomponents/spec/shadow/#dfn-retargeting-algorithm">retargeting algorithm</a>
```
@annevk, is there an equivalent for "retargeting algorithm" in the DOM spec, or is it still best to reference Shadow DOM here?

annevk replied:
All DOM has here is "If target’s root is a shadow-including inclusive ancestor of parent" I think in event dispatch. Perhaps that should be refactored if it needs to be reusable somehow...

TakayoshiKochi replied:
So is referring to Shadow DOM spec here okay?  I followed how activeElement is defined, which uses the same logic.

annevk replied:
I think ideally Shadow DOM disappears so we should figure out what's best here. @hayatoito?

hayatoito replied:
Event dispatch's _retargeting_ in DOM Standard did not need _retargeting_ defined in Shadow DOM spec, _magically_, by using tuples carefully.

Here, it is time to use a retargeting algorithm defined in Shadow DOM from other specifications. This is the first time, I think. If required, define it in DOM Standard at first, then refer to it.

domenic replied:
Apparently this is the second time, if activeElement is doing something similar.

IMO it should be fine to merge this PR with the Shadow DOM reference and then it can be @annevk's responsibility to figure out the right abstraction to put in DOM and change all call sites (this and activeElement) with follow-up PRs. But maybe @annevk disagrees and would rather do everything at once.

annevk replied:
I'm okay with that, but I have limited bandwidth so other people helping out with figuring out the abstractions we need and the PRs would definitely be appreciated.

scheib commented:
I'll review again after previous comments addressed, but looking good to me so far.

TakayoshiKochi commented:
Updated the patch.
Mainly addressed reverting not-in "shadow-including document" to "document".

TakayoshiKochi commented:
@hayatoito FYI

hayatoito commented on index.html:
```diff
@@ -239,9 +239,9 @@
-          the
+          <p>Pointer lock must succeed only if the <a>target</a>'s
+          <a href="https://dom.spec.whatwg.org/#concept-shadow-including-root">shadow-including root</a>
+          is the
```
s/is the/is in the/

hayatoito commented on index.html:
```diff
@@ -275,12 +275,12 @@
           </p>
 
-          <p>If any element (including this one) in the same document
+          <p>If any element (including this one) in the same shadow-including document
```
"in a shadow-including document" is defined in DOM Standard, but strictly speaking, "in the same shadow-including document" is not defined, I think.

How about "If any element (including this one) who shares the same shadow-including root of the target"?

TakayoshiKochi replied:
Assuming the "in the same shadow-including document" [PR#273](https://github.com/whatwg/dom/issues/273) will be accepted, can I leave this part as is?

hayatoito commented on index.html:
```diff
@@ -322,21 +322,12 @@
 
           <p>Initiates an exit from pointer lock state if currently locked
-          to a target in this document,
+          to a target in this shadow-including document,
```
Hmm. The same problem is here.
DOM Standard defines "in _a_ shadow-including document", but we do not have a definition of "a shadow including document", so we can not use "_this_ shadow-including document" here.

annevk replied:
I think the original check is wrong too. You probably want something like "to a target whose shadow-including root is this document".

hayatoito replied:
I recommend to define "A and B are _in the same shadow-including document_ if A and B are in a shadow-including document, and A and B share the same shadow-including root"

annevk replied:
Seems reasonable. Create a PR/issue?

hayatoito replied:
Ops. I did not mean to reply Anne's comment. :) That's a reply to my comment. Anyway, let me (or @TakayoshiKochi) create a PR.

TakayoshiKochi replied:
Even if the "in the same shadow-including document" is defined, I think what @annevk suggested sounds better.

hayatoito commented on index.html:
```diff
@@ -431,7 +440,7 @@
 
       <p>Pointer lock must be exited if the
       <a>target</a> is removed from its document, or the <a>user agent</a>,
```
"removed from its document" must be updated, I think.

TakayoshiKochi replied:
...if the target is _disconnected_, or the ...

look okay?  It seems "disconnected" is not defined.  How about

... if the target is _removed and is no longer in a shadow-including document_, or the ...

?

hayatoito commented on index.html:
```diff
@@ -431,7 +440,7 @@
       <p>Pointer lock must be exited if the
       <a>target</a> is removed from its document, or the <a>user agent</a>,
-      window, or tab loses focus. Moving focus between elements of
+      window, or tab loses focus. Moving focus between elements whose shadow-including root is
```
Could you re-check this sentence? I does not look correct to me.

s/active documents/active document/ ???

TakayoshiKochi replied:
Hmm, my change seems wrong.

The sentence looks meaning that moving focus from one active document to another, and they can be in different browsing context (e.g. main page and iframe).  So the plural form of "active document" should be fine.

So I don't think change is necessary here.

domenic commented on index.html:
```diff
@@ -275,12 +275,12 @@
           <a>pointerlockchange</a> event sent.</p>
 
-          <p>If any element in another document is already locked the request
+          <p>If any element in another shadow-including document is already locked the request
```
This has a similar problem to the above one @hayatoito noted. Although it is probably not a big deal since I think any reader can figure out what it means. But IMO it would be clearer to talk about roots explicitly:

"If any element whose shadow-including root is a different document is already locked, the request must fail..."

domenic commented on index.html:
```diff
@@ -322,21 +322,12 @@
 
           <p>Initiates an exit from pointer lock state if currently locked
-          to a target in this document,
+          to a target whose shadow-including root is this document,
```
Add xref link?

domenic commented on index.html:
```diff
@@ -430,7 +439,8 @@
 
       <p>Pointer lock must be exited if the
-      <a>target</a> is removed from its document, or the <a>user agent</a>,
+      <a>target</a> is removed and is no longer in a shadow-including document,
```
I think this should use https://html.spec.whatwg.org/#node-is-disconnected: "Pointer lock must be exited if the target is [disconnected](https://html.spec.whatwg.org/multipage/infrastructure.html#node-is-disconnected), or the user agent, window, or tab loses focus."

TakayoshiKochi commented:
Updated the patch draft.
Waiting for https://github.com/whatwg/dom/issues/238 to update "in the same shadow-including document" or "connected to the same document".

TakayoshiKochi commented:
Updated the description for `DocumentOrShadowRoot.pointerLockElement`, clarifying the cases whether or not the locked target is in the shadow-including descendant of the context object or not. Thus the spec no longer depends on the "retargeting algorithm" defined in the Shadow DOM spec.

Still waiting https://github.com/whatwg/dom/issues/238 to resolve, but I appreciate if you could review the
updated part of `pointerLockElement` description.

domenic commented on index.html:
```diff
@@ -349,6 +341,43 @@
+           <a href="https://dom.spec.whatwg.org/#concept-shadow-including-descendant">shadow-including descendant</a> of the
+           <a href="http://dom.spec.whatwg.org/#context-object">context object</a>,
+	   return the result of the following algorithm:
+	   <dl>
```
You can reference shadow DOM for this, sorry for confusing you!

If you did want to inline the algorithm, I think it's better to fit with the conventions of this spec, instead of shadow DOM's idiosyncratic CAPITAL LETTER VARIABLES and input/output `<dl>`.

But just referencing shadow DOM's algorith, like you did before, is fine. I just thought it's something that should move to DOM eventually, so I commented on that.

TakayoshiKochi replied:
@hayatoito pointed out in the code review (for Blink) that strictly speaking "retargeting" is originally designed for event retargeting and differs from what `activeElement` does in that the result should be in the same tree as the context object:

> activeElement must be the result of the retargeting algorithm ..., **if the result and the context object are in the same tree**.

So eventually we have to put an algorithm for `activeElement`, `pointerLockElement`, and `fullscreenElement` which slightly differs from retargeting somewhere (in dom spec).

Fixed CAPITAL LETTER VARIABLES, input/output to make it look more natural in this spec.

domenic commented on index.html:
```diff
@@ -430,7 +459,8 @@
 
       <p>Pointer lock must be exited if the
-      <a>target</a> is removed from its document, or the <a>user agent</a>,
+      <a>target</a> is <a href="https://html.spec.whatwg.org/#node-is-disconnected">disconnected</a>,
```
Nit: put the "is " inside the `<a>`.

TakayoshiKochi replied:
Done.

TakayoshiKochi commented:
Updated the patch.
whatwg/dom#238 made me change "in the same shadow-including document"
to "shadow-including root is same".

domenic commented:
LGTM!

scheib commented:
Thank you, LGTM but I will wait for @hayatoito to comment before merging.

hayatoito commented on index.html:
```diff
@@ -349,6 +343,41 @@
+             <li>If <var>target</var> is in the same tree as the
+               <a href="http://dom.spec.whatwg.org/#context-object">context object</a>,
+               let <var>adjusted-target</var> be <var>target</var></li>
+             <li>Otherwise: let <var>adjusted-target</var> be the shadow host which is a
```
Hmm. This definition is not equivalent to the definition which the current `activeElement` is using.
In your definition, there are multiple candidates of shadow hosts which meet the condition. You can not use "the" shadow host here. :(

Could you try to update the condition? I am happy to help you if you need it. :)

TakayoshiKochi replied:
I thought that 
- there should be one shadow host that matches the conditions, so "THE shadow host"
- which meets both of the following conditions (AND condition):
  - "A shadow-including ancestor" of target (because there can be multiple ancestor shadow hosts), and
  - is "in THE same tree" as the context object (this should filter only one in the same tree)

therefore I can use "THE shadow host" here.
I'll ask some native speaker around.

TakayoshiKochi replied:
(talked offline) I understand the condition that can break my assumption :)
I'll update the sentence.

TakayoshiKochi replied:
Updated to return the nearest one in the ancestry.  PTAL.

hayatoito commented on index.html:
```diff
@@ -349,6 +343,42 @@
+             <li>If <var>target</var> is in the same tree as the
+               <a href="http://dom.spec.whatwg.org/#context-object">context object</a>,
+               let <var>adjusted-target</var> be <var>target</var></li>
+             <li>Otherwise: let <var>adjusted-target</var> be the shadow host of <var>target</var> and repeat the following substeps:
```
Unfortunately, you can not use "shadow host of A" if A is not DocumentFragment. :(
DOM Standard defines "A's host" only if A is a DocumentFragment, I think.

Instead, you might want to use "A's root’s host" (if A's root is a shadow root, of course).

TakayoshiKochi replied:
Updated to use "A's root's host" instead of "A's shadow host".

The precondition requires "the target is in the shadow-including descendant of the context object", i.e. if the context's root is not the same as the target's, the target's root must be a shadow root.

TakayoshiKochi commented:
@haytoito PTAL

hayatoito commented on index.html:
```diff
@@ -349,6 +343,42 @@
+         <dd>
+           <dfn title="pointerLockElement"></dfn>
+
+           <p>If the element, which is set as the target for mouse events while the pointer is
```
I recommend you to use a retargeting algorithm as is so that we do not have a lot of specilized retargeting algorithms.
I am afraid that we will have a maintanace burden.

I think your definition works here because it is for DocumentOrShadowRoot, and it is for Pointer locked element.
If one of these assumption breaks, your definition will not work.
I have to review your definition carefully here...
 e.g. The case where the target's root is neither Document nor ShadowRoot. Have you considered this case??

I prefer having one algorithm, and we should refer it from everywhere if we can apply it, instead of customizing it in many places.

TakayoshiKochi replied:
Got the point.

Updated to use retargeting algorithm.
PTAL.

TakayoshiKochi commented:
@hayatoito Ping?

hayatoito commented:
LGTM.
Let me work on upstreaming retargeting algorithm to DOM Standard , https://github.com/whatwg/dom/issues/276.
After that, we can update Pointer Lock specification, again, so that it does not refer to Shadow DOM spec.

TakayoshiKochi commented:
@hayatoito Thanks for the review!

@scheib would you mind merging this?

I'll update the reference to the "retargeting algorithm" (from shadow dom spec to DOM spec)
once whatwg/dom#276 is resolved.

scheib commented on index.html:
```diff
@@ -349,6 +343,34 @@
+           <a href="https://dom.spec.whatwg.org/#concept-shadow-including-descendant">shadow-including descendant</a> of the
+           <a href="http://dom.spec.whatwg.org/#context-object">context object</a>,
+	   returns the result of the
+           <a href="https://https://w3c.github.io/webcomponents/spec/shadow/#dfn-retargeting-algorithm">retargeting algorithm</a>
```
double https

TakayoshiKochi replied:
Fixed, and other instances of using "http:" instead of "https" also fixed.

scheib commented on index.html:
```diff
@@ -275,12 +275,15 @@
-          <a>pointerlockchange</a> event sent.</p>
+          <p>If any element (including this one), whose
+          <a href="https://dom.spec.whatwg.org/#concept-shadow-including-root">shadow-including root</a>
+	  is same as this element's
```
indent

TakayoshiKochi replied:
hard tabs sneaked in.  Fixed.

scheib commented on index.html:
```diff
@@ -275,12 +275,15 @@
+          to this element and a <a>pointerlockchange</a> event sent.</p>
 
-          <p>If any element in another document is already locked the request
+          <p>If any element, whose <a href="https://dom.spec.whatwg.org/#concept-shadow-including-root">shadow-including root</a> is a different document,
```
Wrap long lines please

scheib commented on index.html:
```diff
@@ -349,6 +343,34 @@
+           <a href="https://https://w3c.github.io/webcomponents/spec/shadow/#dfn-retargeting-algorithm">retargeting algorithm</a>
+           with the
+           <a href="http://dom.spec.whatwg.org/#context-object">context object</a> and
+	   the element as input, if the result and the
```
indent

scheib commented on index.html:
```diff
@@ -349,6 +343,34 @@
+           <a href="http://dom.spec.whatwg.org/#context-object">context object</a> and
+	   the element as input, if the result and the
+           <a href="http://dom.spec.whatwg.org/#context-object">context object</a> are in the
+           same tree, otherwise returns null.</p>
```
What situations can cause the retargeting result and the context object to not be in the same tree?

This algorithm is difficult to understand and warrants examples.

TakayoshiKochi replied:
Will add an example.

The situation happens e.g. when 2 shadow hosts exist in the same tree (thus 2 shadow roots, `root1` and `root2`),
and the when pointer locked element is under `root1`.
`root1.pointerLockElement` returns the exact element, `document.pointerLockedElement` returns `root`'s host.
For `root2.pointerLockElement`, if we apply the retargeting algorithm with `root2` and the target, the result will
be `root1`'s host, and is _visible_ from `root2`, but not in or descendant of `root2`'s tree.  Therefore
`root2.pointerLockElement` returns null.

TakayoshiKochi replied:
Oops, the condition "target ... is a shadow-including descendant" already excluded the
possibility that the result of the retargeting algorithm and the context object is in the different trees.
I.e. if we keep the descendant condition, the same tree check won't be necessary.

For the consistency of activeElement algorithm and future reference changes, I'd like to keep the
"run retarget algorithm, then same tree check" part, and would like to drop the descendant
condition.

What do you think @hayatoito?

TakayoshiKochi commented:
@hayatoito I updated the algorithm, dropping the condition to check if the target is shadow-including descendant of the context object or not.  Without the condition check, the case should be covered by the same tree check of retargeting result and context object.  Alternatively, if we keep the shadow-including descendant check, we can drop the same tree check.
I think it is better to drop the shadow-including descendant check to keep in sync with the current activeElement spec for future maintenance cost.

What do you think?

TakayoshiKochi commented:
I've always used git push -f to update the change here, but just in case you want to see
the revision history, see my "shadowdom2" branch:
https://github.com/w3c/pointerlock/compare/gh-pages...TakayoshiKochi:shadowdom2

scheib commented on index.html:
```diff
@@ -349,6 +344,50 @@
+                &lt;/div&gt;
+              &lt;/body&gt;
+          </pre>
+          <p>Note: the example uses fictional <code>shadow-root</code> element to denote a
```
In this section I think that all ids can be used raw, without needing the element type prefixes. E.g.
canvas#canvas1 -> canvas1
div#host1 -> host1

scheib commented:
Thanks for the example. The simplified algorithm is much easier to understand this way as well. I'll wait for hayatoito, but otherwise LGTM.

hayatoito commented:
> I think it is better to drop the shadow-including descendant check to keep in sync with the current activeElement spec for future maintenance cost.

Yeah, you should drop it. Please be consistent with activeElement. That's what I suggested.

LGTM.

TakayoshiKochi commented:
Thanks for all your reviews!

Now I uploaded the final one, fixing https://github.com/w3c/pointerlock/pull/4#discussion_r70138653 to remove typename where ID exists.

hayatoito commented:
@TakayoshiKochi 

Now you can use https://dom.spec.whatwg.org/#retarget. Could you update the spec again?

TakayoshiKochi commented:
Sorry for the delay, #8 is the PR for the update.
